In [3]:
pip install pymongo 


  Using cached dnspython-2.6.1-py3-none-any.whl.metadata (5.8 kB)
   ---------------------------------------- 0.0/773.8 kB ? eta -:--:--
   ---------------------------------------- 773.8/773.8 kB 4.1 MB/s eta 0:00:00
Using cached dnspython-2.6.1-py3-none-any.whl (307 kB)
Note: you may need to restart the kernel to use updated packages.


In [2]:
# Importing package
import pymongo

# Connect to MongoDB
client = pymongo.MongoClient("mongodb://localhost:27017/")

# Accessing the Database
mydb = client["database"]

# Creating Collection
collection = mydb.assignment

# Accessing Collection
collection = mydb["assignment"]

1.Insert the given sample data into a MongoDB collection called 'students'.

In [5]:
# sample data 
students= [
        {"name": "Rohit", "age": 23, "math_score": 85, "physics_score": 90, "city": "New York"},
        {"name": "Eram", "age": 22, "math_score": 78, "physics_score": 75, "city": "Los Angeles"},
        {"name": "Madan", "age": 24, "math_score": 95, "physics_score": 88, "city": "Chicago"},
        {"name": "Uvaish", "age": 21, "math_score": 60, "physics_score": 65, "city": "Houston"},
        {"name": "Neha", "age": 23, "math_score": 72, "physics_score": 80, "city": "Phoenix"}
    ]

# insert many document
collection.insert_many(students)

InsertManyResult([ObjectId('66f00d8a720666f117cfc69f'), ObjectId('66f00d8a720666f117cfc6a0'), ObjectId('66f00d8a720666f117cfc6a1'), ObjectId('66f00d8a720666f117cfc6a2'), ObjectId('66f00d8a720666f117cfc6a3')], acknowledged=True)

discription:The output displays the details of each student in the 'students' collection, including their name, age, math score, physics score, and city. Each student record is printed as a dictionary, along with a unique _id generated by MongoDB.

2.Write a query to find all students who have a 'math_score' greater than 80.

In [4]:
query = {"math_score": {"$gt": 80}}
for students in collection.find(query):
    print(students)

{'_id': ObjectId('66f00989ba9e63a692ddbc44'), 'name': 'Rohit', 'age': 23, 'math_score': 85, 'physics_score': 90, 'city': 'New York'}
{'_id': ObjectId('66f00989ba9e63a692ddbc46'), 'name': 'Madan', 'age': 24, 'math_score': 95, 'physics_score': 88, 'city': 'Chicago'}


discription : This code retrieves and prints student records from the 'students' collection with a math_score greater than 80. It uses a query with the $gt operator to filter results, iterating through the matching documents and printing each one.

3. Write a query to find students whose 'age' is less than 23. 

In [5]:
students_under_23 = collection.find({"age": {"$lt": 23}})
for student in students_under_23:
    print(student) 

{'_id': ObjectId('66f00989ba9e63a692ddbc45'), 'name': 'Eram', 'age': 22, 'math_score': 78, 'physics_score': 75, 'city': 'Los Angeles'}
{'_id': ObjectId('66f00989ba9e63a692ddbc47'), 'name': 'Uvaish', 'age': 21, 'math_score': 60, 'physics_score': 65, 'city': 'Houston'}


discription:The query retrieves Eram (age 22) and Uvaish (age 21) as the students whose ages are less than 23. This confirms that both students meet the specified age criterion.

4.Write a query to return only the 'name' and 'math_score' of all students.

In [6]:
for student in collection.find({}, {"_id": 0, "name": 1,"math_score":1}):
    print(student)

{'name': 'Rohit', 'math_score': 85}
{'name': 'Eram', 'math_score': 78}
{'name': 'Madan', 'math_score': 95}
{'name': 'Uvaish', 'math_score': 60}
{'name': 'Neha', 'math_score': 72}


discription: This code fetches and prints the name and math_score of all students from a MongoDB collection, excluding the _id field. It provides a clear view of each student's key attributes in a concise format.

5.Write a query to find students from the city 'New York'

In [8]:
query = {"city": "New York"}
results = collection.find(query)
for student in results:
    print(student)

{'_id': ObjectId('66f00989ba9e63a692ddbc44'), 'name': 'Rohit', 'age': 23, 'math_score': 85, 'physics_score': 90, 'city': 'New York'}


In [7]:
pipeline = [
    {
        "$group": {
            "_id": {
                "name": "$name",
                "age": "$age",
                "math_score": "$math_score",
                "physics_score": "$physics_score",
                "city": "$city"
            },
            "ids": {"$addToSet": "$_id"},
            "count": {"$sum": 1}
        }
    },
    {"$match": {"count": {"$gt": 1}}}
]

duplicates = list(collection.aggregate(pipeline))

# Step 2: Remove duplicates, keeping one
for duplicate in duplicates:
    collection.delete_many({"_id": {"$in": duplicate["ids"][1:]}})

print("Duplicates removed.")

Duplicates removed.


discription : This code is collection to find all students located in "New York." It then prints each student's complete document, providing insight into their data based on the specified city filter.

6. Write a query to update the 'physics_score' of 'Uvaish' to 85

In [9]:
query = {"name":"Uvaish"}
update={"$set":{"physics_score":85}}
collection.update_one(query,update) 
uvaish_record = collection.find_one({"name": "Uvaish"})
print(uvaish_record)

{'_id': ObjectId('66f00989ba9e63a692ddbc47'), 'name': 'Uvaish', 'age': 21, 'math_score': 60, 'physics_score': 85, 'city': 'Houston'}


discription : The update query sets Uvaish's physics_score to 85. The retrieval confirms that the score has been successfully updated in the database.

7.Write a query to delete the student 'Neha' from the collection

In [10]:
collection.delete_one({"name":"Neha"}) 
remaining_students = collection.find()
for student in remaining_students:
    print(student)

{'_id': ObjectId('66f00989ba9e63a692ddbc44'), 'name': 'Rohit', 'age': 23, 'math_score': 85, 'physics_score': 90, 'city': 'New York'}
{'_id': ObjectId('66f00989ba9e63a692ddbc45'), 'name': 'Eram', 'age': 22, 'math_score': 78, 'physics_score': 75, 'city': 'Los Angeles'}
{'_id': ObjectId('66f00989ba9e63a692ddbc46'), 'name': 'Madan', 'age': 24, 'math_score': 95, 'physics_score': 88, 'city': 'Chicago'}
{'_id': ObjectId('66f00989ba9e63a692ddbc47'), 'name': 'Uvaish', 'age': 21, 'math_score': 60, 'physics_score': 85, 'city': 'Houston'}


discription: indicates that deleted 'Neha' from the students.

8.Write a query to find all students where 'math_score' is between 70 and 90 (inclusive).

In [11]:
query = {"math_score": {"$gt": 70,"$lt":90}}
for students in collection.find(query):
    print(students)

{'_id': ObjectId('66f00989ba9e63a692ddbc44'), 'name': 'Rohit', 'age': 23, 'math_score': 85, 'physics_score': 90, 'city': 'New York'}
{'_id': ObjectId('66f00989ba9e63a692ddbc45'), 'name': 'Eram', 'age': 22, 'math_score': 78, 'physics_score': 75, 'city': 'Los Angeles'}


discription: The output shows student documents, including names, ages, and scores, with multiple entries for 'Rohit' and 'Eram,' suggesting duplicates. Only 'Rohit' has a math score above 80.

9.Write a query to find students whose 'math_score' is greater than 'physics_score'.

In [13]:
pipeline = [
    {
        "$match": {
            "$expr": { "$gt": ["$math_score", "$physics_score"] }  # Compare fields
        }
    }
]

results = collection.aggregate(pipeline)

# Print the output
for student in results:
    print(student)

{'_id': ObjectId('66f00989ba9e63a692ddbc45'), 'name': 'Eram', 'age': 22, 'math_score': 78, 'physics_score': 75, 'city': 'Los Angeles'}
{'_id': ObjectId('66f00989ba9e63a692ddbc46'), 'name': 'Madan', 'age': 24, 'math_score': 95, 'physics_score': 88, 'city': 'Chicago'}


discription:This code finds students, specifically Madan and Eram, whose math_score exceeds their physics_score using MongoDB's aggregation framework. It utilizes the $expr operator for direct field comparisons.





10.Write a query to return students sorted by 'math_score' in descending order

In [14]:
print("\nStodents sorted by math_score(descending) :")
for student in collection.find().sort("math_score",pymongo.DESCENDING):
    print(student) 


Stodents sorted by math_score(descending) :
{'_id': ObjectId('66f00989ba9e63a692ddbc46'), 'name': 'Madan', 'age': 24, 'math_score': 95, 'physics_score': 88, 'city': 'Chicago'}
{'_id': ObjectId('66f00989ba9e63a692ddbc44'), 'name': 'Rohit', 'age': 23, 'math_score': 85, 'physics_score': 90, 'city': 'New York'}
{'_id': ObjectId('66f00989ba9e63a692ddbc45'), 'name': 'Eram', 'age': 22, 'math_score': 78, 'physics_score': 75, 'city': 'Los Angeles'}
{'_id': ObjectId('66f00989ba9e63a692ddbc47'), 'name': 'Uvaish', 'age': 21, 'math_score': 60, 'physics_score': 85, 'city': 'Houston'}


discription:This query retrieves all student records from the assignment collection and orders them by their math_score in descending order, ensuring that the students with the highest scores appear first. The use of sort() allows for efficient data organization based on academic performance.

11. Write a query using the aggregation framework to calculate the average 'math_score' of all students.

In [15]:
pipeline = [
    {
        "$group": {
            "_id": None,  # Grouping all documents together
            "average_math_score": { "$avg": "$math_score" }  # Calculate average
        }
    }
]

results = collection.aggregate(pipeline)
for result in results:
    print("Average Math Score:", result["average_math_score"])

Average Math Score: 79.5


discription: This query calculates the average math_score of all students using the MongoDB aggregation framework. It groups all documents together with _id: None and applies the $avg operator to compute the average score is 79.5.

12.Write a query to group students by 'city' and count the number of students in each city.

In [16]:
pipeline = [
    {
        "$group": {
            "_id": "$city",  # Group by city
            "student_count": { "$sum": 1 }  # Count the number of students in each city
        }
    }
]

# Execute the aggregation pipeline
result = collection.aggregate(pipeline)

# Print the result
for item in result:
    print(f"City: {item['_id']}, Number of Students: {item['student_count']}")

City: New York, Number of Students: 1
City: Chicago, Number of Students: 1
City: Houston, Number of Students: 1
City: Los Angeles, Number of Students: 1


discription: The output of this query displays the number of students grouped by city. For each city, it shows the city name along with the total count of students in that city.

13. Write a query to find students who either have 'math_score' greater than 80 or 'physics_score' greater than 85.

In [17]:
query ={
    "$or":[
        {"math_score": {"$gt": 80}},
        {"math_score": {"$gt": 80}}]}
for students in collection.find(query):
    print(students)

{'_id': ObjectId('66f00989ba9e63a692ddbc44'), 'name': 'Rohit', 'age': 23, 'math_score': 85, 'physics_score': 90, 'city': 'New York'}
{'_id': ObjectId('66f00989ba9e63a692ddbc46'), 'name': 'Madan', 'age': 24, 'math_score': 95, 'physics_score': 88, 'city': 'Chicago'}


discription:This query retrieves students whose math scores are greater than 80. It uses the $or operator, but in this case, both conditions are the same, effectively filtering for students with scores over 80.

14. Write a query to find students whose age is exactly 23 and live in 'Phoenix'.

In [18]:
query = {
    "age": 23,
    "city": "Phoenix"
}

# Execute the query
result = collection.find(query)

# Print the result
for student in result:
    print(student)

discription:The code queries for students aged 23 and living in Phoenix. Since Neha, the only matching student, has been deleted, the query returns no results.

15. Write a query to find students whose 'math_score' is not equal to 85.

In [19]:
query={
    "math_score":{"$ne":85}
}
result=collection.find(query)
for student in result: 
    print(student)

{'_id': ObjectId('66f00989ba9e63a692ddbc45'), 'name': 'Eram', 'age': 22, 'math_score': 78, 'physics_score': 75, 'city': 'Los Angeles'}
{'_id': ObjectId('66f00989ba9e63a692ddbc46'), 'name': 'Madan', 'age': 24, 'math_score': 95, 'physics_score': 88, 'city': 'Chicago'}
{'_id': ObjectId('66f00989ba9e63a692ddbc47'), 'name': 'Uvaish', 'age': 21, 'math_score': 60, 'physics_score': 85, 'city': 'Houston'}


discription:The code finds students whose math_score is not equal to 85, retrieving records for Eram, Madan, and Uvaish. It uses the $ne operator to filter out those with a score of 85.

16. Write a query to find students whose 'name' starts with the letter 'A'.

In [3]:
query={
    "name":{"$regex":"^A"}
}
result = collection.find(query)
for student in result: 
    print(student) 

17. Write a query to find students whose 'city' is neither 'New York' nor 'Los Angeles'.

In [21]:
query = {
    "city": { "$nin": ["New York", "Los Angeles"] }  # $nin is the operator for "not in"
}

results = collection.find(query)
for student in results:
    print(student)

{'_id': ObjectId('66f00989ba9e63a692ddbc46'), 'name': 'Madan', 'age': 24, 'math_score': 95, 'physics_score': 88, 'city': 'Chicago'}
{'_id': ObjectId('66f00989ba9e63a692ddbc47'), 'name': 'Uvaish', 'age': 21, 'math_score': 60, 'physics_score': 85, 'city': 'Houston'}


discription: The query returns students whose city is neither 'New York' nor 'Los Angeles'. In the dataset, this includes Madan (Chicago), Uvaish (Houston).

18. Write a query to update all students with 'age' greater than 22 to add a new field 'graduated': true.

In [22]:
collection.update_many(
    {"age":{"$gt":22}},
    {"$set":{"graduated":True}}
)
updated_students = collection.find()
for student in updated_students:
    print(student)

{'_id': ObjectId('66f00989ba9e63a692ddbc44'), 'name': 'Rohit', 'age': 23, 'math_score': 85, 'physics_score': 90, 'city': 'New York', 'graduated': True}
{'_id': ObjectId('66f00989ba9e63a692ddbc45'), 'name': 'Eram', 'age': 22, 'math_score': 78, 'physics_score': 75, 'city': 'Los Angeles'}
{'_id': ObjectId('66f00989ba9e63a692ddbc46'), 'name': 'Madan', 'age': 24, 'math_score': 95, 'physics_score': 88, 'city': 'Chicago', 'graduated': True}
{'_id': ObjectId('66f00989ba9e63a692ddbc47'), 'name': 'Uvaish', 'age': 21, 'math_score': 60, 'physics_score': 85, 'city': 'Houston'}


discription:The code updates students older than 22, including Rohit and Madan, by adding a graduated field set to true. It then retrieves and prints all student records to show the updated information.

19. Write a query to remove the 'physics_score' field for students who live in 'Chicago'.

In [23]:
query = { "city": "Chicago" }  # Condition for city
update = { "$unset": { "physics_score": "" } }  # Unset the physics_score field

# Perform the update
collection.update_many(query, update)

print("Removed physics_score field for students in Chicago.")

Removed physics_score field for students in Chicago.


discription:The code removes the physics_score field from all students living in Chicago. It then retrieves and prints the updated records for these students, reflecting the changes made.

20. Write a query to find the student with the highest 'math_score'.

In [24]:
highest_math_score_student=collection.find_one(sort=[("math_score",-1)])
print(highest_math_score_student) 

{'_id': ObjectId('66f00989ba9e63a692ddbc46'), 'name': 'Madan', 'age': 24, 'math_score': 95, 'city': 'Chicago', 'graduated': True}


discription: The query identifies Madan as the student with the highest math_score of 95. This result confirms that he achieved the top score among all students in the dataset.